lib/ansible/modules/identity/keycloak/keycloak_scope.py

```diff
@@ -0,0 +1,472 @@
+    if cid is None:
+        if target == 'client':
+            if module.params.get('client_id') is not None:
+                cid = kc.get_client_id(module.params.get('client_id'), realm=realm)
+            else:
+                module.fail_json(msg='When target is client, you need to specify an id or client_id.')
+        else:
+            if module.params.get('name') is not None:
+                cid = kc.get_client_template_id(module.params.get('name'), realm=realm)
+            else:
+                module.fail_json(msg='When target is client-template, you need to specify an id or a name.')
+    if cid is None:
+        module.fail_json(msg='Could not obtain valid client(-template) id to work on.')
+
+    if roletype == 'realm':
+        # in this case, the loop below needs to be run once with a None item
+        scope_clients = [None]
+    if roletype == 'client':
+        scope_clients = list(clientroles.keys())
+        if state == 'exclusive':
+            # when exclusive is set, we want to delete already existing scope mappings regarding all
+            # clients, so add existing mappings' client-names to list of clients to iterate over
+            # (with an empty scope mapping)
+            currentmappings = kc.get_scope_mappings(cid, target=target, realm=realm)
+            if 'clientMappings' in currentmappings:
+                currentmappings = currentmappings['clientMappings']
+            else:
+                currentmappings = dict()
+
+            for mapping in currentmappings.keys():
+                if mapping not in scope_clients:
+                    scope_clients.append(mapping)
+
+    # Let's do some bookkeeping
+    removed_roles = 0
+    updated_roles = 0
+    for scope_client in scope_clients:
+        if scope_client is not None:
+            # If this is a client scope mapping as opposed to a realm scope mapping
+            if scope_client in clientroles:
+                roles = clientroles[scope_client]
+            else:
+                roles = []
+            scope_client_id = kc.get_client_id(scope_client, realm=realm)
+        else:
+            scope_client_id = scope_client
+
+        if roles is None:
+            roles = []
+
+        # obtain current roles
+        current = kc.get_scope_mapping(cid, id_client=scope_client_id, target=target, realm=realm)
+        proposed = deepcopy(current)
+        end_state = []
+
+        to_delete = []
+        if state == 'absent':
+            for role in current:
+                if contains_role(roles, role['name']):
+                    to_delete.append(role)
+                    proposed = [x for x in proposed if x['name'] != role['name']]
+
+        if state == 'exclusive':
+            to_delete = []
+            for role in current:
+                if not contains_role(roles, role['name']):
+                    to_delete.append(role)
+                    proposed = [x for x in proposed if x['name'] != role['name']]
```

Comment:
proposed also gets overwritten for each role in current here. I'm not quite sure what it is intended to look like... if purging is true, would proposed just be any new roles being added? `[new for new in roles if new['name'] not in [current_role['name'] for current_role in current]]`
Or all the playbook-specified roles, whether or not they exist yet? `proposed = list(roles)`

if purging is false it could be something like
```
proposed = deepcopy(current)
proposed.extend([x for x in roles if x not in proposed])
```

Reply:
The idea for proposed is to start with the current state returned by the API and then change it reflecting the changes done/intended by the module -- so if the module intends to remove a role, remove it from proposed as well, etc. In check mode, this give an indication of what the end state would be if the module were executed as configured. In the non-check-mode-case (i.e. the normal running of the module), ideally proposed will match end_state.